### PR TITLE
fix(models): coerce numbers to strings

### PR DESCRIPTION
**Pull request by lengau**
- [ ] Have you followed the guidelines for contributing?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `tox`?

-----

### CODE REVIEW

File: tests/unit/test_models.py [hunk 1]
```diff
@@ -377,13 +389,27 @@ class GrammarValidation(pydantic.BaseModel):
     err = raised.value.errors()
     assert len(err) == 1
     assert err[0]["loc"] == ("x",)
-    assert err[0]["type"] == "value_error"
-    assert err[0]["msg"] == f"Value error, value must be a str: {value!r}"
+    assert err[0]["type"] == "string_type"
+    assert err[0]["msg"] == "Input should be a valid string"
```
Comment (lengau) [inline tests/unit/test_models.py:393]: This change uses the standard pydantic form when the type isn't a dictionary. (The dict type was moved to `test_grammar_errors` below)

File: tests/unit/test_models.py [hunk 2]
```diff
@@ -415,9 +440,9 @@ class GrammarValidation(pydantic.BaseModel):
         )
     err = raised.value.errors()
     assert len(err) == 1
-    assert err[0]["loc"] == ("x",)
-    assert err[0]["type"] == "value_error"
-    assert err[0]["msg"] == "Value error, value must be a str: [35]"
+    assert err[0]["loc"] == ("x", 0, 1)
+    assert err[0]["type"] == "string_type"
+    assert err[0]["msg"] == "Input should be a valid string"
```
Comment (lengau) [inline tests/unit/test_models.py:445]: This is now reproducing the pydantic error as if you'd put `[35]` into a `str` field, but with the correct location.

File: craft_grammar/models.py [hunk 3]
```diff
@@ -64,6 +65,100 @@ def _grammar_append(cls, entry: list[Any], item: Any, info: ValidationInfo) -> N
...
+                                loc=(),
+                                input=item,
+                            ),
+                        ],
+                    )
+                    break
+            else:
+                raise ValueError(  # noqa: TRY004
+                    _format_type_error(type_, input_value),
+                )
```
Comment (mattculler) [inline craft_grammar/models.py:153]: "Flat is better than nested"

```suggestion
            if not issubclass(type_, str):
                raise ValueError(  # noqa: TRY004
                    _format_type_error(type_, input_value),
                )
            if not isinstance(item, dict):
                raise pydantic.ValidationError.from_exception_data(
                    title=f"Grammar[{type_.__name__}]",
                    line_errors=[
                        pydantic_core.InitErrorDetails(
                            type="string_type",
                            loc=(),
                            input=item,
                        ),
                    ],
                )            
            errors.append(
                pydantic_core.InitErrorDetails(
                    type="value_error",
                    loc=(index,),
                    input=item,
                    ctx={
                        "error": ValueError(
                            f"value must be a str or valid grammar dict: {input_value!r}",
                        ),
                    },
                ),
            )
```

**Review by tigarmo — APPROVED** [review]
thanks for the fix! As we discussed the code is getting very hairy but the fix is welcome and any refactoring is *definitely* beyond the scope of this PR ;)

**Review by mr-cal — APPROVED** [review]
Thanks!

File: craft_grammar/models.py (hunk 3 above)
Comment (lengau) [reply craft_grammar/models.py]: Thanks!

**Pull request merged**